storage/block/merged_block_reader.go

```diff
@@ -104,11 +106,23 @@ func (r *dbMergedBlockReader) mergedReader() (xio.SegmentReader, error) {
 		r.readers[i] = nil
 	}
 
-	r.merged = r.encoder.Stream()
+	r.merged = xio.NewBlockReader(r.encoder.Stream(), r.blockStart, r.blockEnd)
 
 	return r.merged, nil
 }
 
+func (r *dbMergedBlockReader) Clone() xio.Reader {
+	return r
```

Comment:
It can't really just return itself as the position into the stream needs to be reset.  You'll need to return `newDatabaseMergedBlockReader(r.blockStart, r.blockEnd, r.streams[0].Clone(), r.streams[1].Clone(), r.opts)` or similar.